src/type_object.rs

```diff
@@ -135,64 +131,117 @@ where
-
-                type_object
-            })
-            .as_ref()
+    pub fn get_or_init<T: PyClass>(&self, py: Python) -> *mut ffi::PyTypeObject {
+        let type_object = *self.value.get_or_init(py, || {
+            let mut type_object = Box::new(ffi::PyTypeObject_INIT);
+            initialize_type_object::<T>(py, T::MODULE, type_object.as_mut()).unwrap_or_else(|e| {
+                e.print(py);
+                panic!("An error occurred while initializing class {}", T::NAME)
+            });
+            Box::into_raw(type_object)
+        });
+
+        // We might want to fill the `tp_dict` with python instances of `T`
+        // itself. In order to do so, we must first initialize the type object
+        // with an empty `tp_dict`: now we can create instances of `T`.
+        //
+        // Then we fill the `tp_dict`. Multiple threads may try to fill it at
+        // the same time, but only one of them will succeed.
+        //
+        // More importantly, if a thread is performing initialization of the
+        // `tp_dict`, it can still request the type object through `get_or_init`,
+        // but the `tp_dict` may appear empty of course.
+
+        if self.tp_dict_filled.get(py).is_some() {
+            // `tp_dict` is already filled: ok.
+            return type_object;
+        }
+
+        {
+            let thread_id = thread::current().id();
+            let mut threads = self.initializing_threads.lock();
+            if threads.contains(&thread_id) {
+                // Reentrant call: just return the type object, even if the
+                // `tp_dict` is not filled yet.
+                return type_object;
+            }
+            threads.push(thread_id);
+        }
+
+        // Pre-compute the class attribute objects: this can temporarily
+        // release the GIL since we're calling into arbitrary user code. It
+        // means that another thread can continue the initialization in the
+        // meantime: at worst, we'll just make a useless computation.
+        let mut items = vec![];
+        for attr in py_class_attributes::<T>() {
+            items.push((attr.name, (attr.meth)(py)));
+        }
+
+        // Now we hold the GIL and we can assume it won't be released until we
+        // return from the function.
+        let result = self.tp_dict_filled.get_or_init(py, move || {
+            let tp_dict = unsafe { (*type_object).tp_dict };
+            let result = initialize_tp_dict(py, tp_dict, items);
+            // See discussion on #982 for why we need this.
+            unsafe { ffi::PyType_Modified(type_object) };
+
+            // Initialization successfully complete, can clear the thread list.
+            // (No further calls to get_or_init() will try to init, on any thread.)
+            *self.initializing_threads.lock() = Vec::new();
+            result
+        });
+
+        if let Err(err) = result {
+            err.clone_ref(py).print(py);
```

Comment:
Why we need `clone` here?

Reply:
We only have a `&Result<(), PyErr>`, and it seems that `PyErr::print` takes `self` as an argument, not `&self`. So I used `clone_ref` (we are in an unlikely error path anyway), but if there is a better way I can change that.

Reply:
Ah, I see, thanks. Then I think we need `clone_ref`.